src/main/scala/org/broadinstitute/hail/variant/Genotype.scala

```diff
@@ -455,41 +451,83 @@ object Genotype {
     new Genotype(gt, ad, dp, gq, pl, flagFakeRef(flags))
   }
 
-  def gen(v: Variant): Gen[Genotype] = {
+  def parseGts(i: Int, j: Int): Int = {
+    val (min, max) = if (i <= j) (i, j) else (j, i)
+    triangle(min) + max
+  }
+
+  def gen(v: Variant): Gen[Gen[Genotype]] = {
     val m = Int.MaxValue / (v.nAlleles + 1)
-    for (gt: Option[Int] <- Gen.option(Gen.choose(0, v.nGenotypes - 1));
-         ad <- Gen.option(Gen.buildableOfN[Array[Int], Int](v.nAlleles,
-           Gen.choose(0, m)));
-         dp <- Gen.option(Gen.choose(0, m));
-         gq <- Gen.option(Gen.choose(0, 10000));
-         pl <- Gen.option(Gen.buildableOfN[Array[Int], Int](v.nGenotypes,
-           Gen.choose(0, m)))) yield {
-      gt.foreach { gtx =>
-        pl.foreach { pla => pla(gtx) = 0 }
-      }
-      pl.foreach { pla =>
-        val m = pla.min
-        var i = 0
-        while (i < pla.length) {
-          pla(i) -= m
-          i += 1
+    Gen.const(
+      for (gt: Option[Int] <- Gen.option(Gen.choose(0, v.nGenotypes - 1));
+           ad <- Gen.option(Gen.buildableOfN[Array[Int], Int](v.nAlleles,
+             Gen.choose(0, m)));
+           dp <- Gen.option(Gen.choose(0, m));
+           gq <- Gen.option(Gen.choose(0, 10000));
+           pl <- Gen.option(Gen.buildableOfN[Array[Int], Int](v.nGenotypes,
+             Gen.choose(0, m)))) yield {
+        gt.foreach { gtx =>
+          pl.foreach { pla => pla(gtx) = 0 }
+        }
+        pl.foreach { pla =>
+          val m = pla.min
+          var i = 0
+          while (i < pla.length) {
+            pla(i) -= m
+            i += 1
+          }
+        }
+        val g = Genotype(gt, ad,
+          dp.map(_ + ad.map(_.sum).getOrElse(0)), gq, pl)
+        g.check(v)
+        g
+      })
+  }
+
+  def genRealistic(v: Variant): Gen[Gen[Genotype]] = {
```

Comment:
I don't think these should be nested generates.  Just use Gen.parameterized, and use the random number generator in the params to cook up frequencies and callRate.